contracts/receivables/BaseCreditPoolReceivable.sol

```diff
@@ -0,0 +1,237 @@
+    mapping(uint256 => ReceivableInfo) public receivableInfoMapping;
+
+    // The status of the receivable.
+    enum Status {
+        Unpaid,
+        Paid,
+        PartiallyPaid
+    }
+
+    // The payment method used to settle the receivable
+    enum PaymentMethod {
+        // Payment is declared by calling the declarePayment function on this contract.
+        // No ERC20 tokens are transferred in this call. The receivable owner is responsible
+        // for transferring the ERC20 tokens to the pool on their own.
+        Declarative,
+        // Payment is made by transferring ERC20 tokens to the pool
+        // using the makePayment function on the BaseCreditPool contract.
+        Payable
+    }
+
+    /**
+     * @dev Emitted when the owner of a receivable calls the declarePayment function
+     * @param from The address of the owner of the receivable
+     * @param to The address of the BaseCreditPool that's paid by the receivable
+     * @param tokenId The ID of the receivable token
+     * @param amount The amount that was declared paid
+     */
+    event PaymentDeclared(
+        address indexed from,
+        address indexed to,
+        uint256 indexed tokenId,
+        uint256 amount
+    );
+
+    /**
+     * @dev Constructor that sets the default admin and minter roles
+     */
+    constructor() ERC721("BaseCreditPoolReceivable", "pREC") {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(MINTER_ROLE, msg.sender);
+    }
+
+    /**
+     * @dev Creates a new receivable token and assigns it to the recipient address
+     * @param recipient The address that will receive the new token
+     * @param baseCreditPool The address of the BaseCreditPool that's paid by the receivable
+     * @param paymentToken The ERC20 token used to pay the receivable
+     * @param receivableAmount The total amount of the receivable
+     * @param maturityDate The date at which the receivable becomes due
+     * @param paymentMethod The payment method used to settle the receivable
+     * @param uri The URI of the metadata associated with the receivable
+     */
+    function safeMint(
+        address recipient,
+        address baseCreditPool,
+        address paymentToken,
+        uint96 receivableAmount,
+        uint64 maturityDate,
+        PaymentMethod paymentMethod,
+        string memory uri
+    ) public onlyRole(MINTER_ROLE) {
+        (address underlyingToken, , , ) = BaseCreditPool(baseCreditPool).getCoreData();
+        require(
+            underlyingToken == paymentToken,
+            "Payment token does not match pool underlying token"
+        );
+
+        uint256 tokenId = _tokenIdCounter.current();
+        _tokenIdCounter.increment();
+        _safeMint(recipient, tokenId);
+
+        ReceivableInfo memory receivableInfo = ReceivableInfo(
+            BaseCreditPool(baseCreditPool),
+            paymentMethod,
+            paymentToken,
+            receivableAmount,
+            0, // Balance
```

Comment:
0 does not seem to be right, should it be receivableAmount?

Reply:
Saw your later implementation, when a payment is declared, you increased balance. You are consistent with your usage of balance, but it seems it is the opposite on the normal use of balance for receivables. For the way how you are implemented, you probably should call it paidAmount.

Reply:
Will rename